src/main/java/com/google/devtools/build/lib/skyframe/SkyframeBuildView.java

```diff
@@ -283,6 +284,10 @@ public void setConfigurations(
     } else {
       String diff = describeConfigurationDifference(configurations, maxDifferencesToShow);
       if (diff != null) {
+        if (failOnAnalysisCacheDiscard) {
+          throw new AnalysisCacheDiscardedException(diff);
```

Comment:
What happens if you alternatively trigger an `eventHandler.handle(Event.error())`?  I think that'd be a lot less of a code diff, at the expense of less control over the exit code?

Reply:
My intention was to stop the build and none of the other exceptions seemed to quite fit? `Event.error()` alone only seemed to print the error to the UI.

Reply:
Oh, I see. I'm not sure offhand which reporting mechanisms trigger which failures modes in this file.

I see `setConfiguration()` is only called by `BuildView.update()`, and the latter throws an `InvalidConfigurationException`. What if we throw that `InvalidConfigurationException` from here and feed its constructor the code at https://github.com/bazelbuild/bazel/blob/01ad0a2d42c95c1525efe7899fc55a15c74c760f/src/main/protobuf/failure_details.proto#L584? 

I think that would produce:

1. a build failure
2. details to understand the specific failure (by reviewing the [build event protocol](https://bazel.build/remote/bep) output) 
3. less new boilerplate code (no need for new exception types or distinct return codes)

Reply:
That works, but I don't see where the detailed exit code in the build configuration gets set in the BEP.

Reply:
Hope you had good holidays - I'm paying attention now.

I'm not sure I understood your question. Does https://github.com/bazelbuild/bazel/blob/b310a8aaf35d73d1310b2a914a459a9f32a04d0d/src/main/java/com/google/devtools/build/lib/buildeventstream/proto/build_event_stream.proto#L543 plus https://github.com/bazelbuild/bazel/blob/ea88c5955edc0ce9c0c7d362225d46d43011bdb5/src/main/java/com/google/devtools/build/lib/buildtool/BuildTool.java#L475-L476 plus https://github.com/bazelbuild/bazel/blob/b310a8aaf35d73d1310b2a914a459a9f32a04d0d/src/main/java/com/google/devtools/build/lib/buildtool/BuildTool.java#L677 (which incorporates the exit code) answer it? Or are you asking something deeper?

I'm fuzzy myself on how exactly build events route to the BEP. I believe the above code links it all together, but I could be wrong.

Reply:
It seems that the `Code` from [`BuildConfiguration`](https://github.com/bazelbuild/bazel/blob/a5c82982b845907fb888cebd3253e8ee13c1fab3/src/main/protobuf/failure_details.proto#L584) gets translated into the [`ExitCode`](https://github.com/bazelbuild/bazel/blob/a5c82982b845907fb888cebd3253e8ee13c1fab3/src/main/java/com/google/devtools/build/lib/util/ExitCode.java#L38), so while it's possible to tell that it exited due to `COMMAND_LINE_ERROR`, without adding a new exit code it doesn't seem possible to see the failure detail from the BEP, as it's not associated with an action here.

That kinda leads me back to preferring the distinct exit code, but seems we can reduce the need for a new exception etc.

Reply:
Ah, I think you're right. I'd like to get some input from people more familiar with this protocol to make sure we're honoring expectations about where we want build failure details to proliferate.

@anakanemison  @justinhorvitz - summary is this PR adds a new analysis phase error. What's the right injection point for reporting, among Bazel exit codes, BEP output, and `failure_details.proto`?

Reply:
Im having an out-of-band chat discussion with others that can hopefully help with my last comment.

The feedback I've gotten so far is pretty strong pushback against adding a new exit code, which I think is fair for a number of reasons. Apparently the issue here is `FailureDetail` does get reported on events like `ActionExecuted` and `TargetComplete`. But not on `BuildFinished`, which I guess is the more appropriate even there. I'm wondering if adding `FailureDetail` to `BuildFinished` would be the right way to unblock us.

Reply:
Are you asking how to propagate a custom exit code? Can you use `AbruptExitException`? That contains the `DetailedExitCode`. Looks like this commit is propagating an exception from `BuildView#update`, which already declares `AbruptExitException`.

Reply:
After more reading of this thread, I think you're trying to see the `DetailedExitCode` in the BEP? I don't know whether my suggestion achieves that. Maybe @michaeledgar could help.

Reply:
@gregestren is the out-of-band chat discussion complete? This seems stalled.

Reply:
Here's further feedback from @michaeledgar copied from another chat channel:

> The issue actually looks like it might be better suited to a different BEP event entirely (https://bazel.build/remote/bep-glossary#aborted). 

> That said, Aborted uses an ancient enum to list abort reasons and doesn't have a FailureDetail, but perhaps it should? These are my thoughts on adding FailureDetail to more events in general:

> 1. In general, adding FailureDetail to other relevant events is great! The semantics just need to be well-defined/reasonable enough to be summarized in the build_event_stream.proto file.
> 2. It does get trickier to achieve #1 at the higher layers of status reporting -- Bazel may see many errors concurrently and will need to pick one. (ActionExecuted is trivial, but even TargetComplete has to pick an error from many: https://github.com/bazelbuild/bazel/blob/7793f7b09e9e49a1988e6d7b993b852781bd1dea/src/main/java/com/google/devtools/build/lib/analysis/TargetCompleteEvent.java#L154-L161
> 3. Aborted might be easier to extend than BuildComplete for that reason, but they should be considered independently. 

@mattem is this enough context for you to work with? Or do you need further clarification or still argue for a different approach?

I realize this is a silly complication for what's a conceptually straightforward change.  But I think it's more broadly good to encourage standard patterns on this front.

Reply:
I poked around more.

While I appreciate the idea of changing the `Aborted` event, there's a distinct `Aborted` for each built target. So `$bazel build //foo //bar` repeats that message, which IMO makes it more awkward to cleanly parse.

What if we add failure details to the `finished` event? 

I prototyped this at https://github.com/bazelbuild/bazel/compare/844b5d2dbc3ff1dda43d78b81a00c6323e4e8115...gregestren:bazel:for_pr_16805?diff=split.

This produces:

```
finished {
  finish_time_millis: 1678842939304
  exit_code {
    name: "COMMAND_LINE_ERROR"
    code: 2
  }
  finish_time {
    seconds: 1678842939
    nanos: 304000000
  }
  failure_detail {
    message: "example message"
    build_configuration {
      code: CONFLICTING_CONFIGURATIONS
    }
  }
}
```

I don't think the concern in the last comment about "picking an error" would really apply in this case.